enhancements/cvo/upgrade_preflight_checks.md

```diff
@@ -0,0 +1,169 @@
+Cluster version operator (CVO) will be running a single job per operator for the preflight checks.
+/bin/preflight must be executable. It can be symbolic link to the operator binary or a script or It can be non existent.
+Users can skip or override the preflight checks with `--force`.
+
+### User Stories
+
+#### Story 1
+
+Some customers would prefer to disable automatic credential provisioning by Cloud credential operator (CCO) and instead manage it themselves.
+However if new permissions are required by the target version, the upgrade could block midway even if the administrator provisioned more powerful credentials.
+With the help of preflight checks CCO can check if they can safely upgrade to a given release before doing the actual upgrade.
+
+### Implementation Details
+
+CVO will run the preflight check from the new operator image to which the current operator image is trying to upgrade.
+Because the current version of operator would not have knowledge of what code changes coming with the new version.
+So its the incoming version's responsibility to figure out if the current image can update safely to it.
+However CVO need to go through signature verification to pass on the target release before we launch preflight checks.
+
+CVO will look for manifests with a specific annotation i.e. openshift.io/preflight-check: "true" and use the manifests to start the preflight jobs.
+
+Some more important points:
+
+* Operators can choose the name and namespace for the job and there will be no restriction from CVO on this.
+* However when CVO will run these jobs it will append some randomized suffix for keeping the jobs unique. Because there can be many preflight jobs running at the same time, so we need a way keep them unique.
+* Preflight checks should not mutate any existing resources as it will make the checks non-idempotent. So it is advisable to give read only access to preflight checks.
+
+
+#### Registering for preflight checks
+
+Operators need to register themselves for a preflight check by defining a Kubernetes job (v1) manifest.
+Preflight checks will be skipped for operator which does not have the manifest.
+
+An example of preflight job manifest
+```
+$ cat 0000_50_cloud-credential-operator_11_preflight-job.yaml
+apiVersion: batch/v1
+kind: Job
+metadata:
+  name: cloud-credential-preflight-checks
+  annotations:
+    openshift.io/preflight-check: "true"
+spec:
+  template:
+    spec:
+      serviceAccountName: openshift-cloud-credential-preflight
+      containers:
+      - name: cloud-credential-operator
+        image: quay.io/openshift/origin-cloud-credential-operator:latest
+          resources:
+            requests:
+               cpu: 10m
+              memory: 150Mi
+        command: ["/bin/preflight",  "foo", "baar"]
+  backoffLimit: 4
+```
+
+
+*Note:*
+To give specific access to preflight jobs, you need to create a new ServiceAccount which the job can use and give the ServiceAccount required RBAC.
```

Comment:
Point out that this RBAC needs to be set up in the running release, not the candidate release which provides the preflight?

Reply:
This should be handled as part of the preflight job. As long as it is idempotent it should be ok.